### PR TITLE
Fix Muon Editing Constraints Causing A Crash

### LINKED ISSUE #29994: Muon Editing Constraints crash
**Original reporter:** Peter Baker / ISIS

### Expected behavior
Editing constraints on parameters in Muon Analysis is handled smoothly.
### Actual behavior
After a simultaneous fit, editing constraints can crash and close workbench
### Steps to reproduce the behavior
- In the Muon Analysis Interface, load 3 consecutive runs e.g. `EMU 62250-62252`
- In the fitting tab, set the fit to simultaneous (and I also changed Run to Group)
- Right-click and add some fit function e.g. `MuonExpDecay`
- Right-click and fix one of the parameters
- Click `Fit`
- Now right-click on the fixed parameter and select `Constraints > Custom`
- Set LowerBound to something, I chose `9`. Hit enter.
- Mantid now hard crashes (closes).
### Platforms affected
I assume all, found on Windows, Workbench 5.1.0, but I can confirm its also a problem on 5.1.1

Steps from Peter:
```
I think this was:
-          Load three consecutive muon runs into Muon Analysis interface
-          Conduct simultaneous fit with all parameters fixed apart from the fields, which were fixed individually
-          Right click on a parameter, choose Constraints -> Custom.

I think it crashed at that point but I might have provided the constraints before it crashed.
```

### PR DESCRIPTION
**Pull request by Matt-Cumber**
**Description of work.**
A read access violation exception was being thrown in a for range loop when attempting to editing constraints on a parameter when fitting. I have altered the loop a bit to have a little more control over what is accessed

**Report to:** Peter Baker / ISIS

**To test:**
Follow original instructions #29994, should no longer crash

Fixes #29994 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/src/FunctionTreeView.cpp [hunk 1]
```diff
@@ -1988,6 +1990,9 @@ void FunctionTreeView::constraintChanged(QtProperty *prop) {
         emit parameterConstraintAdded(functionIndex, constraint);
```
Comment (StephenSmith25) [inline qt/widgets/common/src/FunctionTreeView.cpp:1990]: I'm wondering if we returned after this line we would avoid the crash. I haven't ever looked at this method, but after we've found the constraint property thats changed I wouldn't have thought we'd have to continue looping? I don't think you can change multiple constraints at a time. Maybe you'd have to test it out

Comment (Matt-Cumber) [reply qt/widgets/common/src/FunctionTreeView.cpp:1990]: Ah right, I'll give it a try

Comment (Matt-Cumber) [reply qt/widgets/common/src/FunctionTreeView.cpp:1990]: I can't find a situation where you can change multiple at a time, and I have checked that a return here does also stop a crash as the loop doesn't access memory it shouldn't. So going to change the code back and add a return instead for a cleaner solution. Thanks

**Review by StephenSmith25 — APPROVED** [review]
Looks good to me, it no longer crashes

**Pull request merged**